core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -0,0 +1,251 @@
+     *     or is an OpenWhisk provided image
+     * @param memory memorylimit of the container
+     * @param cpuShares sharefactor for the container
+     * @param environment environment variables to set on the container
+     * @param network network to launch the container in
+     * @param name optional name for the container
+     * @return a container
+     */
+    def create(transid: TransactionId,
+               image: String,
+               userProvidedImage: Boolean = false,
+               memory: ByteSize = 256.MB,
+               cpuShares: Int = 0,
+               environment: Map[String, String] = Map(),
+               network: String = "bridge",
+               name: Option[String] = None)(
+                   implicit docker: DockerApiWithFileAccess, runc: RuncApi, ec: ExecutionContext, log: Logging) = {
+        implicit val tid = transid
+
+        val environmentArgs = (environment + ("SERVICE_IGNORE" -> true.toString)).map {
+            case (key, value) => Seq("-e", s"$key=$value")
+        }.flatten
+
+        val args = Seq(
+            "--cap-drop", "NET_RAW",
+            "--cap-drop", "NET_ADMIN",
+            "--ulimit", "nofile=1024:1024",
+            "--pids-limit", "1024", // OW PR 2119
+            "--cpu-shares", cpuShares.toString,
+            "--memory", s"${memory.toMB}m",
+            "--memory-swap", s"${memory.toMB}m",
+            "--network", network) ++
+            environmentArgs ++
+            name.map(n => Seq("--name", n)).getOrElse(Seq.empty)
+
+        val pulled = if (userProvidedImage) {
+            docker.pull(image).recoverWith {
+                case _ => Future.failed(BlackboxStartupError(s"Failed to pull container image '$image'."))
+            }
+        } else Future.successful(())
+
+        val container = for {
+            _ <- pulled
+            id <- docker.run(image, args)
+            ip <- docker.inspectIPAddress(id, network).andThen {
+                // remove the container immediatly if inspect failed as
+                // we cannot recover that case automatically
+                case Failure(_) => docker.rm(id)
+            }
+        } yield new DockerContainer(id, ip)
+
+        container.recoverWith {
+            case t => if (userProvidedImage) {
+                Future.failed(BlackboxStartupError(t.getMessage))
+            } else {
+                Future.failed(WhiskContainerStartupError(t.getMessage))
+            }
+        }
+    }
+}
+
+/**
+ * Represents a container as run by docker.
+ *
+ * This class contains OpenWhisk specific behavior and as such does not necessarily
+ * use docker commands to achieve the effects needed.
+ *
+ * @constructor
+ * @param id the id of the container
+ * @param ip the ip of the container
+ */
+class DockerContainer(id: ContainerId, ip: ContainerIp)(
+    implicit docker: DockerApiWithFileAccess, runc: RuncApi, ec: ExecutionContext, log: Logging) extends Container with ActionLogDriver {
```

Comment:
i suggest renaming `log` to `logger` or something that more obviously connotes this is the system logger.